sycl/include/sycl/backend.hpp

```diff
@@ -96,7 +96,7 @@ struct BufferInterop {
   GetNativeObjs(const std::vector<ur_native_handle_t> &Handle) {
     ReturnType ReturnValue = 0;
     if (Handle.size()) {
-      ReturnValue = detail::ur::cast<ReturnType>(Handle[0]);
+      ReturnValue = (ReturnType)(Handle[0]);
```

Comment:
Should this be a C++ `static_cast` instead of a C-style cast?

Reply:
Static cast breaks for some types unfortunately

Reply:
@hdelan, I am concerned, I assume that failing reinterpred_cast here could be a sign of real problem. Could you please help me to clarify what exact problem is solved here?

In description you mentioned that the problem is conversion from unsigned long to unsigned long long. From headers I see that ur_native_handle_t is uintptr_t. 
Is my understanding correct that the problem is conversion of **pointer** to unsigned long to **pointer** of unsigned long long? 
If it is correct than the problem should be windows specific because on x64 systems long has the same size as long long on linux and has smaller size on windows https://learn.microsoft.com/en-us/cpp/cpp/data-type-ranges?view=msvc-170. And if this point correct than why do we treat conversion of unsigned long* to unsigned long long* as valid here?

Reply:
The `reinterpret_cast` is from an `unsigned long` to an `unsigned long long`, not from the ptr types.

```
 /home/hugh/llvm/build/bin/../include/sycl/detail/ur.hpp:97:10: error: reinterpret_cast from 'unsigned long' to 'unsigned long long' is not allowed
  97 |   return reinterpret_cast<To>(value);
       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The problem is that `Handle[0]` is a `uintptr_t` which on my linux system is aliased to an `unsigned long`, whereas `backend_return_t<ext_oneapi_cuda, buffer<DataT, Dimensions, AllocatorT>>` is a `CUdeviceptr` which we alias to a `unsigned long long` since `__LP64__` is defined for my Linux system. See here https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/detail/backend_traits_cuda.hpp#L32

This is also how `CUdeviceptr` is defined within the CUDA driver:

See in `cuda.h`
```c++
255- * CUdeviceptr is defined as an unsigned integer type whose size matches the size of a pointer on the target platform.
256- */
257-#if defined(_WIN64) || defined(__LP64__)
258:typedef unsigned long long CUdeviceptr_v2;
259-#else
260:typedef unsigned int CUdeviceptr_v2;
261-#endif
262:typedef CUdeviceptr_v2 CUdeviceptr;

Reply:
I see and double checked, static_cast allows that conversion but could not be used since not working for some other types (as you said above).
thanks for clarification

Reply:
Thanks @KseniyaTikhomirova are you happy to approve the patch?